cmd/snapd-generator/main.c

```diff
@@ -30,30 +31,168 @@
 #include "../libsnap-confine-private/mountinfo.h"
 #include "../libsnap-confine-private/string-utils.h"
 
-static sc_mountinfo_entry *find_root_mountinfo(sc_mountinfo * mounts)
+static sc_mountinfo_entry *find_dir_mountinfo(const char *mnt_dir)
 {
+	// Load /proc/1/mountinfo so that we can inspect the root filesystem.
+	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
+	if (!mounts) {
+		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+		return NULL;
+	}
+
 	sc_mountinfo_entry *cur, *root = NULL;
 	for (cur = sc_first_mountinfo_entry(mounts); cur != NULL;
 	     cur = sc_next_mountinfo_entry(cur)) {
 		// Look for the mount info entry for the root file-system.
-		if (sc_streq("/", cur->mount_dir)) {
+		if (sc_streq(mnt_dir, cur->mount_dir)) {
 			root = cur;
+			break;
 		}
 	}
 	return root;
 }
 
-static int ensure_root_fs_shared(const char *normal_dir)
+#define SYSTEMD_GENERATORS_SVC_DIR     "/run/systemd/generator"
+#define SYSTEMD_GENERATORS_SYSINIT_DIR SYSTEMD_GENERATORS_SVC_DIR "/sysinit.target.wants"
+
+static int create_early_mount(const char *normal_dir,
+			      const char *what,
+			      const char *unit_name)
 {
-	// Load /proc/1/mountinfo so that we can inspect the root filesystem.
-	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
-	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
-	if (!mounts) {
-		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+	// Construct the file name for a new systemd mount unit.
+	char fname[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(fname, sizeof fname,
+			 "%s/%s.mount",  normal_dir, unit_name);
+
+	// Open the mount unit and write the contents.
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "w");
+	if (!f) {
+		fprintf(stderr, "cannot write to %s: %m\n", fname);
 		return 1;
 	}
+	fprintf(f, "[Unit]\n");
+	fprintf(f, "Description=Early mount of kernel drivers tree\n");
+	fprintf(f, "DefaultDependencies=no\n");
+	fprintf(f, "After=systemd-remount-fs.service\n");
+	fprintf(f, "Before=sysinit.target\n");
+	fprintf(f, "Before=systemd-udevd.service systemd-modules-load.service\n");
+	fprintf(f, "Before=umount.target\n");
+	fprintf(f, "Conflicts=umount.target\n");
+	fprintf(f, "\n");
+	fprintf(f, "[Mount]\n");
```

Comment:
Maybe I am missing something. The manpage of `systemd.mount` says that `Where` is mandatory.

Reply:
Done